syntax = "proto3";

package scanner.proto;

// Database metadata messages
message DatabaseDescriptor {
  message Job {
    int32 id = 1;
    string name = 2;
  }

  message Table {
    int32 id = 1;
    string name = 2;
  }

  int32 next_job_id = 1;
  int32 next_table_id = 2;
  repeated Job jobs = 3;
  repeated Table tables = 4;
}

enum DeviceType {
  CPU = 0;
  GPU = 1;
}

enum ImageEncodingType {
  JPEG = 0;
  PNG = 1;
  BMP = 2;
  RAW = 3;
}

enum ImageColorSpace {
  Gray = 0;
  RGB = 1;
  RGBA = 2;
}

enum ColumnType {
  Other = 0;
  Video = 1;
  Image = 2;
}

enum FrameType {
  U8 = 0;
  F32 = 1;
  F64 = 2;
}

message Column {
  int32 id = 1;
  string name = 2;
  ColumnType type = 3;
}

message VideoDescriptor {
  enum VideoCodecType {
    H264 = 0;
    RAW = 1;
  }

  enum VideoChromaFormat {
    Monochrome = 0;
    YUV_420 = 1;
    YUV_422 = 2;
    YUV_444 = 3;
  }

  int32 table_id = 1;
  int32 column_id = 2;
  int32 item_id = 3;

  int64 frames = 4;
  int32 width = 5;
  int32 height = 6;
  int32 channels = 14;
  FrameType frame_type = 13;

  VideoCodecType codec_type = 7;
  VideoChromaFormat chroma_format = 8;
  int32 time_base_num = 15;
  int32 time_base_denom = 16;

  int64 num_encoded_videos = 17;
  repeated int64 frames_per_video = 18;
  repeated int64 keyframes_per_video = 19;
  repeated int64 size_per_video = 20;

  repeated int64 keyframe_positions = 9 [packed=true];
  repeated int64 keyframe_timestamps = 10 [packed=true];
  repeated int64 keyframe_byte_offsets = 11 [packed=true];
  bytes metadata_packets = 12;
}

message ImageFormatGroupDescriptor {
  int32 id = 1;
  ImageEncodingType encoding_type = 2;
  ImageColorSpace color_space = 3;
  int32 width = 4;
  int32 height = 5;
  int32 num_images = 7;
  repeated int64 compressed_sizes = 6 [packed=true];
}

message TableDescriptor {
  int32 id = 1;
  string name = 2;
  repeated Column columns = 3;
  // @brief the last row in each item of this table
  repeated int64 end_rows = 4;
  int32 job_id = 6;
  int64 timestamp = 7;
}

// Task set messages
message TableSample {
  string table_name = 1;
  repeated string column_names = 2;
  string sampling_function = 3;
  bytes sampling_args = 4;
}

message Task {
  string output_table_name = 2;
  repeated TableSample samples = 3;
}

message OpInput {
  int32 op_index = 1;
  repeated string columns = 2;
}

message Op {
  string name = 1;
  repeated OpInput inputs = 2;
  DeviceType device_type = 3;
  bytes kernel_args = 4;
  repeated int32 stencil = 5;
  int32 batch = 6;
  int32 warmup = 7;
}

message OutputColumnCompression {
  string codec = 1;
  map<string, string> options = 2;
}

message TaskSet {
  repeated Task tasks = 1;
  repeated Op ops = 2;
  repeated OutputColumnCompression compression = 3;
}

message JobDescriptor {
  int32 id = 1;
  string name = 2;
  int32 io_item_size = 3;
  int32 work_item_size = 4;
  int32 num_nodes = 5;
  repeated Task tasks = 6;
  repeated Column columns = 7;
}

// Interal messages
message DecodeArgs {
  int32 width = 4;
  int32 height = 5;
  int64 start_keyframe = 6;
  int64 end_keyframe = 7;
  repeated int64 keyframes = 1;
  repeated int64 keyframe_byte_offsets = 2;
  repeated int64 valid_frames = 3;
  int64 encoded_video = 8;
  int64 encoded_video_size = 9;
}

message ImageDecodeArgs {
  int32 warmup_count = 1;
  int32 rows_from_start = 2;
  ImageEncodingType encoding_type = 3;
  ImageColorSpace color_space = 4;
  repeated int64 compressed_sizes = 5 [packed=true];
  repeated int64 valid_images = 6;
}

message LoadSample {
  int32 table_id = 1;
  repeated int32 column_ids = 2;
  int64 warmup_size = 3;
  repeated int64 rows = 4 [packed=true];
}

message LoadWorkEntry {
  int32 job_index = 1;
  int32 task_index = 2;
  repeated LoadSample samples = 3;
  repeated int64 work_item_sizes = 4;
}

message MemoryPoolConfig {
  message Pool {
    bool use_pool = 1;
    int64 free_space = 2;
  }

  bool pinned_cpu = 1;
  Pool cpu = 3;
  Pool gpu = 4;
}

message CollectionDescriptor {
  repeated string tables = 1;
  int32 job_id = 2;
}

message CollectionsDescriptor {
  repeated int64 ids = 1;
  repeated string names = 2;
}

message FrameInfo {
  repeated int32 shape = 1;
  int32 type = 2;
}

message MachineParameters {
  int32 num_cpus = 1;
  int32 num_load_workers = 2;
  int32 num_save_workers = 3;
  repeated int32 gpu_ids = 4;
}

message IOItem {
  // @brief the output table id
  int32 table_id = 1;
  // @brief the unique id for this item in the table
  int64 item_id = 2;
  // @brief the first row in this item
  int64 start_row = 3;
  // @brief the row after the last row in this item
  int64 end_row = 4;
}

// Sampler args

message AllSamplerArgs {
  int64 sample_size = 1;
  int64 warmup_size = 2;
}

message StridedRangeSamplerArgs {
  int64 stride = 1;
  repeated int64 warmup_starts = 2;
  repeated int64 starts = 3;
  repeated int64 ends = 4;
}

message GatherSamplerArgs {
  message Sample {
    repeated int64 warmup_rows = 1 [packed=true];
    repeated int64 rows = 2 [packed=true];
  }

  repeated Sample samples = 1;
}

message PythonArgs {
  bytes py_args = 1;
}
